Further transliteration of ASL semantics rules and start of transliteration of ASL typing rules

jalglave opened the pull request:
hi @HadrienRenaud please could you look at this first attempt at transliterating LHS expressions? In particular Rule.LETuple but all the rest too.
Thanks in advance,
Jade

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,78 @@ Evaluation of Left-Hand-Side Expressions
   {3 Rule.LELocalVar}
-  <description> 
+  [le] denotes a variable [x] and all of the following applies:
+  - [new_env] is a local environment; 
```
```suggestion
  - [x] is locally bound in [new_env]; 
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,78 @@ Evaluation of Left-Hand-Side Expressions
-  <description> 
-  <minimal example>
+  [le] denotes a variable [x] and all of the following applies:
+  - [new_env] is a global environment; 
```
```suggestion
  - [x] is globally bound in [env]; 
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,78 @@ Evaluation of Left-Hand-Side Expressions
 
   {3 Rule.LESlice}
-  <description> 
+  [le] denotes an expression [e_bv] sliced as per [slices] and all of the following applies:
+  - The evaluation of [e_bv] under [env] is [v_bv];
+  - The evaluation of [slices] under [env] is [positions];
+  - [v] is the value read in [v_bv] from [positions];
+  - [new_env] is [env] where [le] has been assigned the value [v].
```
This is beginning to be tricky. Here is a suggestion, but I'm not sure that it is understandable:

```suggestion
  [le] denotes a left-hand-side expression [le_bv] sliced as per [slices] and all of the following applies:
  - [le_bv] has a corresponding right-hand-side expression [e_bv];
  - The evaluation of [e_bv] under [env] is [v_bv];
  - The evaluation of [slices] under [env] is [positions];
  - [new_v_bv] is [v_bv] where the positions [positions] have been updated to [v]
  - [new_env] is [env] where [le_bv] has been assigned the value [new_v_bv].
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,78 @@ Evaluation of Left-Hand-Side Expressions
 
   {3 Rule.LESetArray}
-  <description> 
+  [le] denotes an array [e_array] and an index [e_index] and all of the following applies:
+  - The evaluation of [e_array] under [env] is [v_array];
+  - The evaluation of [e_index] under [env] is [v_index];
+  - [v] is the value found at the index [v_index] of [v_array];
+  - [new_env] is [env] where [le] has been assigned the value [v].
```
```suggestion
  [le] denotes an array [le_array] and an index [e_index] and all of the following applies:
  - [le_array] has a corresponding right-hand-side expression [e_array];
  - The evaluation of [e_array] under [env] is [v_array];
  - The evaluation of [e_index] under [env] is [v_index];
  - [new_v_array] is [v_array] where the value at index [v_index] has been updated to [v];
  - [new_env] is [env] where [le] has been assigned the value [new_v_array].
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,78 @@ Evaluation of Left-Hand-Side Expressions
 
   {3 Rule.LESetField}
-  <description> 
+  [le] denotes a field name [field_name] in a vector [e_vec] and all of the following applies:
+  - The evaluation of [e_vec] under [env] is [v_vec];
+  - [v] is the value mapped by [field_name] in [v_vec];
+  - [new_env] is [env] where [le] has been assigned the value [v].
```
```suggestion
  [le] denotes a field name [field_name] in a vector [le_vec] and all of the following applies:
  - [le_vec] has a corresponding right-hand-side expression [e_vec]
  - The evaluation of [e_vec] under [env] is [v_vec];
  - [new_v_vec] is [v_vec] where the field [field_name] has been updated to [v]
  - [new_env] is [env] where [le] has been assigned the value [new_v_vec].
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,78 @@ Evaluation of Left-Hand-Side Expressions
 
-  {3 Rule.LETupleUnpack}
-  <description> 
+  {3 Rule.LETuple}
+  [le] denotes a list of expression [e_list] and all of the following applies:
+  - the evaluation of [e_list] in [env] is [v_list];
+  - [v] is the vector built from [v_list];
+  - [new_env] is [env] where [le] has been assigned the value [v].
```
```suggestion
  {3 Rule.LETuple}
  [le] denotes a list of expression [le_list] and [new_env] is [env] where each left-hand-side expression [sub_le] has been assigned the value at the corresponding index in [v].
```

HadrienRenaud reviewed (commented):
It's a bit harder to convey the precise meaning of the code, thus those are only suggestions and you might find better ways of expressing what the code means.

jalglave commented:
thanks @HadrienRenaud I think my latest commit addresses your comments, and modifies the code to use the same names. Please could you read through again, and when you suggest names to use in the transliteration, indicate what they correspond to in the code? For example at the moment there is no "sub_le" in the code of LETuple so I can't use that name in the transliteration.

jalglave commented:
@HadrienRenaud please could you take a look at this start on patterns. In particular Rule.PMask but all the others too.
Thanks in advance,
Jade

jalglave commented:
hi @HadrienRenaud please could you take a look at the draft for statements? Especially throws but all the rest too. 
Thanks in advance,
Jade

jalglave commented:
hi @HadrienRenaud please could you take a look at the draft for local declarations? Especially the uninitialised ones but all the rest too.
Thanks in advance,
Jade

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -321,47 +324,81 @@ Evaluation of Left-Hand-Side Expressions
 
   {3 Rule.LESlice}
-  <description> 
+  [le] denotes a left-hand-side expression sliced as per [slices] and
+  all of the following applies:
+  - The right-hand-side expression corresponding to [le] is [e_bv];
+  - The evaluation of [re_bv] under [env] is [rv_bv];
+  - The evaluation of [slices] under [env] is [positions];
+  - [v] is [rv_bv] where the positions [positions] have been updated to [v]
+  - [new_env] is [env] where [le] has been assigned the value [v].
```
```suggestion
  [le] denotes a left-hand-side expression [le_bv] sliced as per [slices] and
  all of the following applies:
  - The right-hand-side expression corresponding to [le_bv] is [re_bv];
  - The evaluation of [re_bv] under [env] is [v_bv];
  - The evaluation of [slices] under [env] is [positions];
  - [new_v_bv] is [v_bv] where the positions [positions] have been updated to [v]
  - [new_env] is [env] where [le_bv] has been assigned the value [new_v_bv].
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -473,80 +543,130 @@ Evaluation of Statements
+  All of the following applies:
+  - [s] is a return statement;
+  - [s] gives a list of expressions [es];
+  - [cont] is [env] modified after evaluation of the list [es] under [env].
```
```suggestion
  - [vs] is the piece-wise evaluation of [es] under [env]
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -473,80 +543,130 @@ Evaluation of Statements
+  All of the following applies:
+  - [s] is a declaration;
+  - [s] gives [(_ldk, ldi, Some e)];
+  - [cont] is [env] modified after evaluation of the local declaration [s ldi env (Some m)].
```
Why `s` in the end?

```suggestion
  - [cont] is [env] modified after evaluation of the local declaration [ldi env (Some m)].
```

HadrienRenaud commented on asllib/ASLSemanticsReference.mld:
```diff
@@ -473,80 +543,130 @@ Evaluation of Statements
+  All of the following applies:
+  - [s] is a declaration;
+  - [s] gives [(_dlk, ldi, None)];
+  - [cont] is [env] modified after evaluation of the local declaration [s ldi env None].
```
Idem
```suggestion
  - [cont] is [env] modified after evaluation of the local declaration [ldi env None].
```

HadrienRenaud reviewed (commented):
Hi Jade,

There has been a few mismatch on recursive evaluation in left-expressions.

For patterns, maybe we want to be precise and say "in the environment [env]" each time? Otherwise sounds good to me.

jalglave commented:
Thanks @HadrienRenaud I think my last commit addresses your comments. Please could you check and also make a suggestion for Rule.PMask? We'll need to double-check that we use the same names in the code and in the spec as mentioned in a previous comment.
Thanks,
Jade

HadrienRenaud commented:
Hi @jalglave 

Here is a draft for PMask:
```
All the following applies:
- [le] is a mask [m]
- [v] is a bitvector that matches the mask [m]
```

jalglave commented:
hi @HadrienRenaud please could you double-check the latest commits (which aim to render our earlier live conversation)? Also please consider adding examples throughout, and making names match in the code and in the spec.
Thanks in advance,
Jade

HadrienRenaud commented:
Hi @maranget, I've pushed the edits I wanted to make, could you review this PR and see it merged please?
